File: Controller.py
import socket
import binascii
import time
import keyboard

CAMERA_IP = 'IP_ADDRESS'
CAMERA_PORT = xxxxx

def send_command_tcp(command): # tcp initialising
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            sock.settimeout(10)
            sock.connect((CAMERA_IP, CAMERA_PORT))
            print(f"Connected to {CAMERA_IP}:{CAMERA_PORT}")

            sock.send(command)
            print(f"Command sent: {command.hex()}")
            time.sleep(1)

            response = sock.recv(4096)
            if not response:
                print("No response received.")
                return None

            print(f"Response received: {response.hex()}")
            return response

    except socket.timeout:
        print("Socket timed out while waiting for a response.")
    except ConnectionRefusedError:
        print("Connection refused by the camera. Check the IP and port.")
    except socket.error as e:
        print(f"Socket error occurred: {e}")
    except Exception as e:
        print(f"An unexpected error occurred while sending command: {e}")
    return None

def parse_firmware_version(response):
    try:
        response_hex = binascii.hexlify(response).decode('ascii')
        print(f"Raw response hex: {response_hex}")

        firmware_version_bytes = response[8:12]  #example only, standard
        firmware_version = ''.join(format(b, '02x') for b in firmware_version_bytes)

        try:
            firmware_version_str = bytes.fromhex(firmware_version).decode('ascii')
        except (ValueError, UnicodeDecodeError):
            firmware_version_str = firmware_version

        return firmware_version_str

    except IndexError as e:
        print(f"Index error while parsing response: {e}")
    except ValueError as e:
        print(f"Value error while converting data: {e}")
    except Exception as e:
        print(f"An unexpected error occurred while parsing firmware version: {e}")
    return None

def parse_camera_status(response):
    try:
        response_hex = binascii.hexlify(response).decode('ascii')
        print(f"Raw response hex: {response_hex}")

        status_bytes = response[8:12]  #example only, standard
        status = ''.join(format(b, '02x') for b in status_bytes)

        try:
            status_str = bytes.fromhex(status).decode('ascii')
        except (ValueError, UnicodeDecodeError):
            status_str = status

        return status_str

    except IndexError as e:
        print(f"Index error while parsing response: {e}")
    except ValueError as e:
        print(f"Value error while converting data: {e}")
    except Exception as e:
        print(f"An unexpected error occurred while parsing camera status: {e}")
    return None

def request_firmware_version():
    command = bytes.fromhex("55 66 01 00 00 00 00 01 64 c4")
    response = send_command_tcp(command)
    if response:
        firmware_version = parse_firmware_version(response)
        if firmware_version:
            print(f"Firmware Version: {firmware_version}")
        else:
            print("Failed to parse firmware version.")
    else:
        print("Failed to receive a response from the camera.")

def auto_zoom_in():
    command = bytes.fromhex("55 66 01 01 00 00 00 05 01 8d 64")
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Camera is Zooming in: {status}")
        else:
            print("Failed to parse camera status.")
    else:
        print("Failed to receive a response from the camera.")

def auto_zoom_out():
    command = bytes.fromhex("55 66 01 01 00 00 00 05 FF 5c 6a")
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Camera is Zooming out: {status}")
        else:
            print("Failed to parse camera status.")
    else:
        print("Failed to receive a response from the camera.")

def Center():
    command = bytes.fromhex("55 66 01 01 00 00 00 08 01 d1 12")
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Centered: {status}")
        else:
            print("Failed to parse camera status.")
    else:
        print("Failed to receive a response from the camera.")

def Camera_Status(): #this is not for camera status, update needed
    command = bytes.fromhex("55 66 01 02 00 00 00 07 64 64 3d cf") #this command is a test command for yaw and pitch for full length
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Camera status: {status}")
        else:
            print("Failed to parse camera status.")
    else:
        print("Failed to receive a response from the camera.")

def Lock_mode():
    command = bytes.fromhex("55 66 01 01 00 00 00 0c 03 57 fe")
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Lock mode activated: {status}")
        else:
            print("Failed to parse camera status.")
    else:
        print("Failed to receive a response from the camera.")

def Follow_mode():
    command = bytes.fromhex("55 66 01 01 00 00 00 0c 04 b0 8e")
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Follow mode activated: {status}")
        else:
            print("Failed to parse camera status.")
    else:
        print("Failed to receive a response from the camera.")

def Gimbal_Rotation():
    command = bytes.fromhex("55 66 01 04 00 00 00 0e 00 00 ff a6 3b 11")
    response = send_command_tcp(command)
    if response:
        status = parse_camera_status(response)
        if status:
            print(f"Gimbal rotation response: {status}")
        else:
            print("Failed to parse gimbal rotation status.")
    else:
        print("Failed to receive a response from the camera.")

def save_response_to_file(response, filename="camera_response.log"):
    with open(filename, 'a') as file:
        file.write(f"{time.ctime()}: {response.hex()}\n")
    print(f"Response saved to {filename}")

if __name__ == "__main__":
    print("Press 'f' to request firmware version.")
    print("Press 'z' to auto zoom in.")
    print("Press 'x' to auto zoom out.")
    print("Press 'c' to center the gimbal.")
    print("Press 's' to get camera status.")
    print("Press 'l' to activate lock mode.")
    print("Press 'o' to activate follow mode.")
    print("Press 'r' to rotate the gimbal.")

    while True:
        if keyboard.is_pressed('f'):
            request_firmware_version()
            time.sleep(1)
        elif keyboard.is_pressed('z'):
            auto_zoom_in()
        elif keyboard.is_pressed('x'):
            auto_zoom_out()
        elif keyboard.is_pressed('c'):
            Center()
        elif keyboard.is_pressed('s'):
            Camera_Status()
        elif keyboard.is_pressed('l'):
            Lock_mode()
        elif keyboard.is_pressed('o'):
            Follow_mode()
        elif keyboard.is_pressed('r'): 
            Gimbal_Rotation()

File: Controller.cpp
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <string>
#include <sstream>
#include <iomanip>
#include <thread>
#include <chrono>

#pragma comment(lib, "Ws2_32.lib")

const std::string CAMERA_IP = "IP_ADDRESS";
const int CAMERA_PORT = xxxxx;

void printHex(const char* data, size_t length) {
    for (size_t i = 0; i < length; ++i) {
        std::cout << std::hex << std::setw(2) << std::setfill('0') << (unsigned int)(unsigned char)data[i];
    }
    std::cout << std::endl;
}

std::string bytesToHex(const char* bytes, size_t length) {
    std::ostringstream oss;
    for (size_t i = 0; i < length; ++i) {
        oss << std::hex << std::setw(2) << std::setfill('0') << (unsigned int)(unsigned char)bytes[i];
    }
    return oss.str();
}

std::string sendCommandTCP(const std::string& command) {
    WSADATA wsaData;
    SOCKET ConnectSocket = INVALID_SOCKET;
    struct sockaddr_in clientService;
    std::string response;

    // winsock initialise
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        std::cerr << "WSAStartup failed" << std::endl;
        return response;
    }

    // socket connection
    ConnectSocket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (ConnectSocket == INVALID_SOCKET) {
        std::cerr << "Error at socket(): " << WSAGetLastError() << std::endl;
        WSACleanup();
        return response;
    }

    // TCP connection
    clientService.sin_family = AF_INET;
    clientService.sin_port = htons(CAMERA_PORT);
    if (inet_pton(AF_INET, CAMERA_IP.c_str(), &(clientService.sin_addr)) <= 0) {
        std::cerr << "Invalid IP address" << std::endl;
        closesocket(ConnectSocket);
        WSACleanup();
        return response;
    }

    //  server coonection
    if (connect(ConnectSocket, (SOCKADDR*)&clientService, sizeof(clientService)) == SOCKET_ERROR) {
        std::cerr << "Failed to connect: " << WSAGetLastError() << std::endl;
        closesocket(ConnectSocket);
        WSACleanup();
        return response;
    }

    // transmit data
    int bytesSent = send(ConnectSocket, command.c_str(), command.size(), 0);
    if (bytesSent == SOCKET_ERROR) {
        std::cerr << "Failed to send command: " << WSAGetLastError() << std::endl;
        closesocket(ConnectSocket);
        WSACleanup();
        return response;
    }

    // receive data
    const int MAX_RECVBUF_SIZE = 4096;
    char recvbuf[MAX_RECVBUF_SIZE];
    int recvbuflen = MAX_RECVBUF_SIZE;

    int bytesReceived = recv(ConnectSocket, recvbuf, recvbuflen, 0);
    if (bytesReceived > 0) {
        response.assign(recvbuf, bytesReceived);
        std::cout << "Received response: ";
        printHex(response.c_str(), response.size());
    }
    else if (bytesReceived == 0) {
        std::cerr << "Connection closed" << std::endl;
    }
    else {
        std::cerr << "recv failed: " << WSAGetLastError() << std::endl;
    }

    // socket cleaner
    closesocket(ConnectSocket);
    WSACleanup();

    return response;
}

std::string parseFirmwareVersion(const std::string& response) {
    if (response.size() < 12) return "Invalid response";
    return bytesToHex(response.c_str() + 8, 4);
}

std::string parseCameraStatus(const std::string& response) {
    if (response.size() < 12) return "Invalid response";
    return bytesToHex(response.c_str() + 8, 4);
}

void requestFirmwareVersion() {
    std::string command = "\x55\x66\x01\x00\x00\x00\x00\x01\x64\xc4";
    std::string response = sendCommandTCP(command);
    std::string firmwareVersion = parseFirmwareVersion(response);
    std::cout << "Firmware Version: " << firmwareVersion << std::endl;
}

void autoZoomIn() {
    std::string command = "\x55\x66\x01\x01\x00\x00\x00\x05\x01\x8d\x64";
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Camera is Zooming in: " << status << std::endl;
}

void autoZoomOut() {
    std::string command = "\x55\x66\x01\x01\x00\x00\x00\x05\xFF\x5c\x6a";
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Camera is Zooming out: " << status << std::endl;
}

void center() {
    std::string command = "\x55\x66\x01\x01\x00\x00\x00\x08\x01\xd1\x12";
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Centered: " << status << std::endl;
}

void cameraStatus() {
    std::string command = "\x55\x66\x01\x02\x00\x00\x00\x07\x64\x64\x3d\xcf"; // Note that this is for Yaw and Pitch at full length
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Camera status: " << status << std::endl;
}

void lockMode() {
    std::string command = "\x55\x66\x01\x01\x00\x00\x00\x0c\x03\x57\xfe";
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Lock mode activated: " << status << std::endl;
}

void followMode() {
    std::string command = "\x55\x66\x01\x01\x00\x00\x00\x0c\x04\xb0\x8e";
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Follow mode activated: " << status << std::endl;
}

void gimbalRotation() {
    std::string command = "\x55\x66\x01\x04\x00\x00\x00\x0e\x00\x00\xff\xa6\x3b\x11";
    std::string response = sendCommandTCP(command);
    std::string status = parseCameraStatus(response);
    std::cout << "Gimbal rotation response: " << status << std::endl;
}

int main() {
    std::cout << "Press 'f' to request firmware version." << std::endl;
    std::cout << "Press 'z' to auto zoom in." << std::endl;
    std::cout << "Press 'x' to auto zoom out." << std::endl;
    std::cout << "Press 'c' to center the gimbal." << std::endl;
    std::cout << "Press 's' to get camera status." << std::endl;
    std::cout << "Press 'l' to activate lock mode." << std::endl;
    std::cout << "Press 'o' to activate follow mode." << std::endl;
    std::cout << "Press 'r' to rotate the gimbal." << std::endl;
    std::cout << "Press 'q' to quit the program." << std::endl;

    while (true) {
        if (GetAsyncKeyState('F') & 0x8000) {
            requestFirmwareVersion();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('Z') & 0x8000) {
            autoZoomIn();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('X') & 0x8000) {
            autoZoomOut();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('C') & 0x8000) {
            center();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('S') & 0x8000) {
            cameraStatus();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('L') & 0x8000) {
            lockMode();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('O') & 0x8000) {
            followMode();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('R') & 0x8000) {
            gimbalRotation();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
        else if (GetAsyncKeyState('Q') & 0x8000) {
            std::cout << "Quitting program." << std::endl;
            break;
        }
    }

    return 0;
}
